### ISSUE TITLE
[Dialog] opening lazy loaded content not currently possible

### ISSUE THREAD

**Issue by spock123** [angular/components#11074]
#### Bug, feature request, or proposal:

Feature request

#### What is the expected behavior?

Hey guys, would it be possible to develop some kind of mechanism that allows for loading routes as dialog content?

Right now, if I want to open a dialog using md-dialog, I have to pass an instance of the content component. This means that the content component needs to be in the same module (or in the global app module).

There are many use cases for this, for example:

2 lazy loaded modules:

A: show list of items, navigate, filter etc
B: show detail view of an item

If I want to show details from the list in module A inside an md-dialog, I cannot do this because the details view component will be unknown to the calling component inside another module.

As it is now, I have to put all components into the same module, so the entire idea of lazy loading my applications goes out the window. The details component could be using all kinds of heavy dependencies that I don't want to load unless the user really requests a details view.

To summarize:

- can material dialog open an url as content of a dialog?

Cheers

**Comment by devversion** [comment]
That seems like a great feature request to me. My thinking is that the the dialog or portal should just also support `ComponentFactory` instances as those then can be lazily loaded through dynamic imports (like with routing; I think string module URLs are not the best long-term approach).

I'd love to see this working:

```ts
// Lazily load the dialog component. Note: For VE compatibility,
// import from `.ngfactory` and append `NgFactory` to the export name.
const contentFactory = (await import('./lazy/dialog-content')).MyDialogContent;

// Open the dialog using the given factory.
this.dialog.open(contentFactory)
```

**Comment by devversion** [comment]
Appendix: Apparently just loading a component dynamically is not sufficient. At the current time where View Engine is still supported, and NgModule's are still a thing, one would need to load the component through the module like this:

```ts
const [moduleFactoryExports, componentExports] = await Promise.all([
  // Note: Uses still factory for backwards compat w/ View Engine.
  import(moduleFactoryFile),
  import(componentFactoryFile),
]);

const moduleFactory: NgModuleFactory<any> = moduleFactoryExports[`${module.name}NgFactory`];
const componentType: Type<any> = componentExports[componentName];
const lazyComponent = moduleFactory.create(injector)
  .componentFactoryResolver.resolveComponentFactory(componentType);

this.dialog.open(lazyComponent);
```

It's actually not as simple as just having an URL for the router because we'd need to find a specific component, while lazy routing modules provide the components through the injector. We could have something similar for the dialog, but ideally with Ivy and no remote scoping, it should be possible to just import the actual component as each component is self-contained (with used directives, pipes etc.; no need for modules)

**Comment by devversion** [comment]
It looks like the first snippet actually works in Ivy as metadata is encapsulated in the component and it is a self-contained factory. Closing! Please let us know if this does not work.
